### PR TITLE
Recipe App - Group project

### PR DESCRIPTION
**Pull request by initseis**
## Description
In this branch we:

- Followed the layout of the wireframes provided.
- Added a Login page and registration page:
    - It was built with Devise.
- Added a Food list:
    - It displays a list of food as in the wireframe.
    - It leads to the form that allows users to add new food.
- Added a Recipies list:
    - It displays a list of recipes created by the logged-in user as in the wireframe.
    - It leads to recipe details.
    - If the user is the owner of the recipe, allow user to delete it.
- Added a Public recipe list:
    - It displays a list of all public recipes ordered by newest as in the wireframe.
    - It leads to recipe details.
- Added Recipe details:
    - If the recipe is public or the user is the owner of the recipe, displays the recipe details as in the wireframe.
    - If the user is the owner of the recipe, leads to the form that allows user to add new food.
- Added a General shopping list view:
    - It shows the list of food that is missing for all recipes of the logged-in user.
    - It counts the total food items and total price of the missing food.
- Made sure there are no N+1 queries happening.
- Created a navigation menu that allows users to open all of the pages you created.

## How Has This Been Tested?
- This project has integration testings with Capybara
  - To check the tests run `rspec rspec/* --format doc`

### CODE REVIEW

File: app/controllers/recipes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  before_action :set_user_recipes, only: %i[show update destroy]
+  before_action :set_user_recipe, only: %i[show update]
+
+  def index
+    @recipes = current_user.recipes
+  end
+
+  def public_recipes
+    @recipes = Recipe.where("public = 'true' ")
+  end
```
Comment (sinansevgi) [inline app/controllers/recipes_controller.rb:11]: - [ ] You should avoid making SQL queries from the controller. Your controllers should only redirect data that they get from models. You should always make SQL queries from the model.

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  def general_shopping_list
+    @foods = current_user.foods
+    @total_price = 0
+    @foods.each do |food|
+      @total_price += food.price * food.food_quantity
```
Comment (sinansevgi) [inline app/controllers/recipes_controller.rb:17]: - [ ] The controllers should be concerned with HTTP verbs. (thin controller rule) By that, I mean that a controller’s responsibility should be focused on web requests. Other calculations should be handled on the model.

File: app/models/food.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Food < ApplicationRecord
+  belongs_to :user
+  has_many :recipe_foods, dependent: :destroy
+
+  def food_quantity
+    quantity = 0
+    recipe_foods.each do |food|
+      quantity += food.quantity
+    end
```
Comment (sinansevgi) [inline app/models/food.rb:9]: - [ ] The controllers should be concerned with HTTP verbs. (thin controller rule) By that, I mean that a controller’s responsibility should be focused on web requests. Other calculations should be handled on the model.

File: app/views/foods/create.html.erb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+<h1>Foods#create</h1>
+<p>Find me in app/views/foods/create.html.erb</p>
```
Comment (sinansevgi) [inline app/views/foods/create.html.erb:2]: [Optional]
- [ ] Please remove unused views.

**Review by sinansevgi — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
------

Hi @sevinchek and @the-catalystmc, 

Good job so far!
Database and columns configured correctly :heavy_check_mark: 
HTML is clean :heavy_check_mark: 
There are some issues that you still need to work on to go further but you are almost there!

 ## Required Changes ♻️
_Please check the inline comments for my suggestions_ :arrow_down:

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Happy Hacking! :computer: 

-----

You can reply to this conversation if you have any questions or need elaboration. :wave: 
Sinan Sevgi

------

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/food.rb (hunk 3 above)
Comment (the-catalystmc) [reply app/models/food.rb:9]: @sinansevgi Hey man, I don't really understand this change. The calculations here are inside the model for food. Isn't this what you are suggesting?

Comment (sinansevgi) [reply app/models/food.rb:9]: You are calculating quantity, you should do it in your model.

**Comment by sinansevgi** [comment]
- [ ] You should have unit and Integration tests for your models and controllers. Please be sure all of your controllers and models have tests

**Review by descholar-ceo — APPROVED** [review]
Hi @the-catalystmc , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**